ESQL: Support reading points from doc-values for STATS ST_CENTROID

craigtaverner opened the pull request:
Support `st_centroid` aggregation on `geo_point` and `cartesian_point` data. All aggregations will combined the original data into new data, and as such we can load from doc-values for performance. This feature is both the first spatial aggregation for ES|QL as well as support for loading points from doc-values instead of from source, but only for spatial aggregations.

Since ES|QL aggregations (ie the `STATS` command) are split into two phases, an initial aggregation on the data nodes, and the final aggregation on the coordinator node, this allows us to only need to support doc-values on the data nodes. The data flow is as follows:

* Data node uses FieldExtractExec to tell the BlockLoaderContext that the blocks should be loaded with `forStats==true`, and the GeoPoint and cartesian Point field types will detect this boolean and return `LongBlock` loaders.
* The `SpatialAggregateFunction` on the data node will be notified that it should support `LongBlock` input, and will decode the long into `(double x, double y)` before aggregating, and producing intermediate results (four `DoubleBLock` for xSum, xComp, ySum, yCom and one `LongBlock` for count).
* The `SpatialAggregateFunction` on the coordinator node only receives intermediate results, and will never receive actual points, so does not care what block type was originally used, and will produce the final centroid in WKB format.

This design has a few advantages over the original plan of supporting both WKB and encoded longs in the full stack:
* Only one backing type for each explicit type (no need to bring back the `LongBlock` support we used in 8.12)
* The decision to switch to `LongBlock` in the data loading from the index can be made very late, at the `LocalPhysicalPlanOptimizer` stage, isolating it to a very small set of changes, and also only on the data nodes at the data loading part.

Note that this solution will not support the use of doc-values in `ST_CENTROID` if some other `STATS` command is run first, because only the first `STATS` command will be run on the data nodes. For example:

```
FROM airports
| STATS count = COUNT(*) BY scalerank, location
| LIMIT 10
| STATS centroid=ST_CENTROID(location), count=SUM(count)
```

In this case the `ST_CENTROID` will use the source values for the calculation. We have decided to favour the simpler solution of only doing local physical planning for doc-values because support of doc-values in the coordinator node requires a much more complex solution with full-stack support of both source and doc-values versions of the point types.

Fixes #104656 104656

craigtaverner committed:
Add initial structure for ST_CENTROID

craigtaverner committed:
Revert "Revert stab at implementing forStats for doc-values vs source"

This reverts commit cfc4341bf4bf139f203671e58a5fc62140d859de.

craigtaverner committed:
Refined csv-spect tests with st_centroid

craigtaverner committed:
Spotless disagrees with intellij

craigtaverner committed:
Fixes after reverting fieldmapper code to test GeoPointFieldMapper

craigtaverner committed:
Get GeoPointFieldMapperTests working again after enabling doc-values reading

craigtaverner committed:
Simplify after rebase on main

In particular, field-mappers that do not need to know about fields can have simpler calls.

craigtaverner committed:
Support local physical planning of forStats attributes for spatial aggregations

craigtaverner committed:
Get st_centroid aggregation working on doc-values

We changed it to produce BytesRef, so we don't (yet) need any doc-values types.

craigtaverner committed:
Create both DocValues and SourceValues versions of st_centroid

craigtaverner committed:
Support physical planning of DocValues and SourceValues SpatialCentroid

craigtaverner committed:
Improve test for physical planning of DocValues in SpatialCentroid

craigtaverner committed:
Fixed show functions for st_centroid

craigtaverner committed:
More st_centroid tests with mv_expand

To test single and multi-value centroids

craigtaverner committed:
Fix st_centroid from point literals

The blocks contained BytesRef byte[] with multiple values, and we were ignoring the offsets when decoding, so decoding the first value over and over instead of decoding the subsequent values.

craigtaverner committed:
Teach CsvTests to handle spatial types alternative loading from doc-values

Spatial GEO_POINT and CARTESIAN_POINT load from doc-values in some cases. If the physical planner has planned for this, we need the CsvTests to also take that into account, changing the type of the point field from BytesRefBlock to LongBlock.

craigtaverner committed:
Fixed failing NodeSubclassTests

Required making the new constructor public and enabling Set as a valid parameter in the test framework.

elasticsearchmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

craigtaverner committed:
More complex st_centroid tests and fixed bug with multiple aggs

When there were multiple agregations in the same STATS, we were inadvertently re-ordering them, causing the wrong Blocks to be fed to the wrong aggregator in the coordinator node.

craigtaverner committed:
Update docs/changelog/104218.yaml

craigtaverner committed:
Fix automatically generated changelog file

craigtaverner committed:
Fixed failing test

The nodes can now sometimes be Set, which is also a Collection, but not a List, and therefor never can be a subset of the children.

craigtaverner committed:
More tests covering more combinations including MV_EXPAND and grouping

craigtaverner committed:
Added cartesian st_centroid with grouping test

We could not add MV_EXPAND tests since the cartesian data does not have multi-value columns, but the geo_point tests are sufficient for this since they share the same code.

craigtaverner committed:
Reduce flaky tests by sorting results

craigtaverner committed:
Reduce flaky tests by sorting results

craigtaverner committed:
Added tests for stats on stats to ensure planner coped

craigtaverner committed:
Add unit tests to ensure doc-values in query planning complex cases

craigtaverner commented on server/src/test/java/org/elasticsearch/index/mapper/GeoPointFieldMapperTests.java:
```diff
@@ -705,13 +721,18 @@ protected IngestScriptSupport ingestScriptSupport() {
-    private boolean useDocValues = false;
+    @Override
+    protected Function<Object, Object> loadBlockExpected() {
+        throw new IllegalStateException("Should never reach here, call loadBlockExpected(boolean) instead");
```
Message refers to wrong method signature

craigtaverner replied:
Done

craigtaverner commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/SpatialCentroidCartesianPointSourceValuesAggregator.java:
```diff
@@ -0,0 +1,56 @@
+
+/**
+ * This aggregator calculates the centroid of a set of cartesian points.
+ * It is assumes that the cartesian points are encoded as longs.
```
Should say WKB here.

craigtaverner replied:
Done

craigtaverner commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/SpatialCentroidGeoPointSourceValuesAggregator.java:
```diff
@@ -0,0 +1,55 @@
+import org.elasticsearch.geometry.utils.WellKnownBinary;
+
+/**
+ * This aggregator calculates the centroid of a set of geo points. It is assumes that the geo points are encoded as longs.
```
Should say WKB here.

craigtaverner replied:
Done

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/PlannerUtils.java:
```diff
@@ -211,6 +211,14 @@ public static ElementType toSortableElementType(DataType dataType) {
      * Map QL's {@link DataType} to the compute engine's {@link ElementType}.
      */
     public static ElementType toElementType(DataType dataType) {
+        // TODO: Remove this method and update all calls to pass the extra boolean
```
With the latest design, the use of `forStats==true` is limited to the local physical plan, so most of the time it is completely correct to set this to false. Should we bother changing all the callers to this method when the case where it is needed is limited to a single location?

craigtaverner commented on x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/index/mapper/PointFieldMapperTests.java:
```diff
@@ -428,8 +428,10 @@ protected IngestScriptSupport ingestScriptSupport() {
-        // Currently ESQL support for cartesian_point is limited to source values
-        return false;
+    protected BlockReaderSupport getSupportedReaders(MapperService mapper, String loaderFieldName) {
+        // TODO: Support testing both reading from source as well as reading from doc-values
```
The problem here is that the entire test class `PointFieldMapperTests` has not been setup for synthetic source tests, which `GeoPointFieldMapperTests` has, and this support is needed for the relevant tests. We can investigate adding this.

craigtaverner replied:
Done - https://github.com/elastic/elasticsearch/issues/104504

craigtaverner commented on benchmarks/src/main/java/org/elasticsearch/benchmark/compute/operator/ValuesSourceReaderBenchmark.java:
```diff
@@ -189,6 +189,11 @@ public String indexName() {
                 }
 
+                @Override
+                public boolean forStats() {
```
We've considered alternative names, like `forDisplay` (opposite of for stats), but every choice has pros and cons, so I've left it like this for now, since this most closely matches current needs.

If we start using this for non-stats use cases, like spatial predicates that benefit from doc-values, we can change the name of the method at that point.

nik9000 replied:
Predicate and top-n run on the results of functions consuming the points will want it.

craigtaverner commented on x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/Methods.java:
```diff
@@ -47,7 +50,7 @@
 public class Methods {
     static ExecutableElement findRequiredMethod(TypeElement declarationType, String[] names, Predicate<ExecutableElement> filter) {
-        ExecutableElement result = findMethod(declarationType, names, filter);
+        ExecutableElement result = findMethod(names, filter, declarationType, superClassOf(declarationType));
```
Added support for inheritence in finding the static methods used in aggregations, because, for example, the SpatialCentroid can be configured in four different ways (geo vs cartesian, and doc-values vs source values). Used only single-level inheritance for simplicity.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/SpatialCentroid.java:
```diff
@@ -0,0 +1,92 @@
+        if (useDocValues) {
+            // When the points are read as doc-values (eg. from the index), feed them into the doc-values aggregator
+            if (type == EsqlDataTypes.GEO_POINT) {
+                return new SpatialCentroidGeoPointDocValuesAggregatorFunctionSupplier(inputChannels);
```
Look at the changes in `AggregateMapper` to see how we supported these combinations of generated function names.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizer.java:
```diff
@@ -427,4 +434,43 @@ public static boolean isPushableFieldAttribute(Expression exp, Predicate<FieldAt
+        protected PhysicalPlan rule(AggregateExec aggregate) {
+            var foundAttribute = new Holder<Attribute>(null);
+
+            PhysicalPlan plan = aggregate.transformDown(UnaryExec.class, exec -> {
```
This is the heart of the solution for figuring out when to use doc-values and when to use source values. It only runs on the local node, so we only consider doc-values on the local node, and it re-writes spatial aggregations to mark fields as doc-values so the right spatial aggregation functions are executed, as well as finding the fields involved, and re-writing the FieldExtractExec to load that field as doc-values from the index.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/physical/FieldExtractExec.java:
```diff
@@ -13,17 +13,25 @@
+        this.forStats = forStats;
+    }
+
+    public FieldExtractExec(Source source, PhysicalPlan child, List<Attribute> attributesToExtract) {
```
Since the re-writing to use `forStats` is done only on the local node, this knowledge does not need to be serialised across the cluster. It is therefor not required to update the transport-version, and this older constructor remains the main constructor for the entire stack. If we decide at some point to plan this knowledge in the logical planner, we would need to serialize this, and update the transport version too. At this point we see no need to take that step.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/AggregateMapper.java:
```diff
@@ -131,50 +137,62 @@ private List<IntermediateStateDesc> getNonNull(AggDef aggDef) {
             types = List.of(""); // no extra type distinction
+        } else if (clazz == SpatialCentroid.class) {
+            types = SPATIAL;
+            extraConfigs = List.of("SourceValues", "DocValues");
```
This is where the additional combination of function names is generated. So far only spatial aggregations need this, but the idea can be used by others if there is ever value.

nik9000 replied:
Could you explain what this is about in more detail? I'd kind of imagined the `withForStats` would be enough to enable the optimization and the mapper would just call the single function. I imagine I'm missing some part of how the linkage works though.

craigtaverner replied:
Take a look at `SpatialCentroid.supplier()`. There are four possible combinations, based on two spatial types `GEO_POINT` and `CARTESIAN_POINT`, and two loading types `DocValues` and `SourceValues`. The code generator needs to generate functions for all four combinations. The `AggregateMapper` deals with many possible combinations, but not this specific set of `DocValues` and `SourceValues`, so I added these using the same pattern I see here for other combinations. Without this fix the classloader cannot find the classes to load for the four combinations.

nik9000 replied:
> Without this fix the classloader cannot find the classes to load for the four combinations.

That's the trick. I suppose we're using some kind of trick to get these linked up. Scalars just use booleans for it. I'd kind of prefer that. But now isn't the time to rework that.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/AggregateMapper.java:
```diff
@@ -131,50 +137,62 @@ private List<IntermediateStateDesc> getNonNull(AggDef aggDef) {
             types = NUMERIC;
         } else if (clazz == Count.class) {
             types = List.of(""); // no extra type distinction
+        } else if (clazz == SpatialCentroid.class) {
```
Perhaps check for `SpatialAggregateFunction` instead, since all spatial aggregations will need this.

craigtaverner reviewed (commented):
Did my own review, with a few suggested changes as well as some explanatory comments.

craigtaverner committed:
Some minor updates from code review

costin commented:
Craig, leaving a quick comment that I'll get to review this PR however due to its size but it will take some time.

craigtaverner committed:
Fixes after rebase on main

luigidellaquila reviewed (commented):
Thanks @craigtaverner 

I'll review it in detail, but I'd like to leave here some of the outcomes of our off-line conversations, in particular regarding the conceptual approach, to make sure we are all on the same page.

The optimization rule that decides how to load the geo values (`SpatialFromDocValuesExtraction`) is in `LocalPhysicalPlanOptimizer`. At that level we only have the portion of the plan that is executed locally, ie. we don't have all the information to understand if the values will be used for display or for aggs.

In particular, a query with a pipeline breaker before STATS won't be optimized.

A practical example is: get the centroid of the top 10 airports:

```
FROM airports_web 
| SORT scalerank desc 
| LIMIT 10 
| STATS centroid=ST_CENTROID(location)
```

For this query, the  SORT + LIMIT part is a pipeline breaker, the STATS will be executed on the coordinator (the local node won't even know about that STATS command), and the optimization won't kick in.

We have to be aware that it's a sub-optimal solution, but we can decide to live with it for now (or forever), especially if the cost of calculating it on the coordinator, propagating the information and retrieving the data in different formats is particularly expensive.

We also have to consider that today we have a limited set of commands, and in practice we have only two pipeline breakers, but in the future we'll have more, so the situation described above will become more and more likely to happen.

It's something we can decide to change in the future, but it will come with a cost for backward compatibility

iverase commented:
@luigidellaquila I think that's ok, note that we are computing the centroid over 10 documents so it is ok to generate doc values on the fly. More over your example is perfect for the optimization proposed [here](https://github.com/elastic/elasticsearch/issues/103636), merging function into the loader. 

Moving that function to the loader. Imagine the centroid is over a geo_shape field, and it contains polygons with trillion of vertices, if we can push the centroid to the loader, we will only need to move the centroid around instead of having to send those enormous shapes to the coordinator.

nik9000 commented:
Interesting about the pipelines breaking operations that might break the STATS detection. I *think* on the theoretical level it's always ok. Like, the breaker will have to reduce the size of the data coming out. The important bit is that we load the shape after the breaker, at least on the local node. And topn is fine with that now I believe. It'll do the local top n on the key and `doc` and then load the fields it needs and ship them to the coordinating top-n. I suppose if the shape is involved in the top-n we do want to use the doc value - but we already should be doing that. Stuff like `FROM foo | SORT MV_COUNT(point) | LIMIT 5000 | STATS whatever` should be able to use the doc values. And, well, in this case `MV_COUNT` is a decent candidate to fuse into the loader too. But that's a problem for another time.

nik9000 commented on test/framework/src/main/java/org/elasticsearch/index/mapper/MapperTestCase.java:
```diff
@@ -1244,60 +1248,99 @@ public final void testBlockLoaderFromColumnReaderWithSyntheticSource() throws IO
-    protected boolean supportsColumnAtATimeReader(MapperService mapper, MappedFieldType ft) {
-        return ft.hasDocValues();
+    /**
+     *  Should the doc-values reader be used instead of reading from source.
```
The javadoc looks out of date.

craigtaverner replied:
Updated, and removed the rowStrideReader boolean, since that was always true. I had added that expecting to need it in the spatial tests, but did not use it.

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_point_doc_values

craigtaverner committed:
Get correct error message on unsupported geo_shape for st_centroid

craigtaverner committed:
Refined point vs shape differences after merging main

craigtaverner committed:
Added basic docs

elasticsearchmachine commented:
Pinging @elastic/es-docs (Team:Docs)

craigtaverner committed:
Delete docs/changelog/104218.yaml

craigtaverner committed:
Revert "Delete docs/changelog/104218.yaml"

This reverts commit 4bc596a442cc4d95c827cdfbb6617f7b9399238d.

craigtaverner committed:
Fixed broken docs tag link

costin reviewed (approved):
This PR looks good! It's tight which made it easy to review - I focused on the QL side and planner.
The only issue I saw was the use of docValues boolean inside SpatialAggregateFunction - it's an extraction concern that should not be on the logical declaration since it's not user defined. It doesn't seem used so I would just remove it - but maybe I've missed anything.

Otherwise LGTM. I'll let Nik comment on the extraction bit.

P.S. On such a large PR of course I had nit comments on naming and such  :)

craigtaverner committed:
Simplify BlockReaderSupport in MapperTestCase from code review

craigtaverner commented on modules/mapper-extras/src/test/java/org/elasticsearch/index/mapper/extras/MatchOnlyTextFieldMapperTests.java:
```diff
@@ -265,7 +265,7 @@ protected IngestScriptSupport ingestScriptSupport() {
 
     @Override
-    protected Function<Object, Object> loadBlockExpected(MapperService mapper, String loaderFieldName) {
+    protected Function<Object, Object> loadBlockExpected() {
```
I'm not sure I follow? I just reverted this method to the version before Luigi's PR. Neither his change nor mine changed the return type.

costin replied:
Just saying that it _might_ be helpful to expose the type in the method signature `protected Function<Object, BytesRef>`.
it's used in testing so it might not matter in the end.

craigtaverner committed:
Moved spatial aggregations into a sub-package

luigidellaquila commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizer.java:
```diff
@@ -427,4 +434,43 @@ public static boolean isPushableFieldAttribute(Expression exp, Predicate<FieldAt
+                            var changed = as.replaceChild(af.withDocValues());
+                            changedAggregates = true;
+                            if (af.field() instanceof Attribute fieldAttribute) {
+                                foundAttribute.set(fieldAttribute);
```
Just a double-check: are you taking this recent change into consideration? https://github.com/elastic/elasticsearch/pull/104387
(I don't have a real use case where this could make a big difference, just wondering if you checked it)

craigtaverner replied:
Looking at that now. It turns out that if we nest a function inside the `st_centroid` then we won't load from doc-values. The queries work, but the optimization is not activated. I'm busy debugging that now to see if there is a simple solution, but this could also be a followup PR to enabled this case.

craigtaverner replied:
I did need to make a small change to the local physical planner to get all tests to pass with nested functions. They are not using the doc-values optimization, but at least the queries run with nested fields.

luigidellaquila reviewed (approved):
Thank you very much @craigtaverner, LGTM!

I just left a comment, not sure it's really important but maybe worth checking

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_point_doc_values

craigtaverner commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/CentroidPointAggregator.java:
```diff
@@ -0,0 +1,278 @@
+ * 2.0.
+ */
+
+package org.elasticsearch.compute.aggregation;
```
Done, but this does involve yet another small hack top AggregateMapper to support lookup of classes in sub-packages.

craigtaverner commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/SpatialCentroidCartesianPointDocValuesAggregator.java:
```diff
@@ -0,0 +1,55 @@
+ * 2.0.
+ */
+
+package org.elasticsearch.compute.aggregation;
```
Moved to `spatial` sub-package.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/EsqlFunctionRegistry.java:
```diff
@@ -169,6 +170,8 @@ private FunctionDefinition[][] functions() {
                 def(DateParse.class, DateParse::new, "date_parse"),
                 def(DateTrunc.class, DateTrunc::new, "date_trunc"),
                 def(Now.class, Now::new, "now") },
+            // spatial
+            new FunctionDefinition[] { def(SpatialCentroid.class, SpatialCentroid::new, "st_centroid") },
```
I choose this location as directly after the `// date` section, since that is the only other cases of a type-specific set of additional functions. Seems like the right location to me. If we move it to the end, should we move `// date` to the end too?

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/SpatialCentroid.java:
```diff
@@ -0,0 +1,93 @@
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+package org.elasticsearch.xpack.esql.expression.function.aggregate;
```
Moved to `spatial` sub-package.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizer.java:
```diff
@@ -427,4 +434,43 @@ public static boolean isPushableFieldAttribute(Expression exp, Predicate<FieldAt
         return false;
     }
 
+    private static class SpatialFromDocValuesExtraction extends OptimizerRule<AggregateExec> {
```
Done

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizer.java:
```diff
@@ -427,4 +434,43 @@ public static boolean isPushableFieldAttribute(Expression exp, Predicate<FieldAt
+                            changedAggregates = true;
+                            if (af.field() instanceof Attribute fieldAttribute) {
+                                foundAttribute.set(fieldAttribute);
+                            } else {
+                                throw new RuntimeException("Expected field attribute, got " + af.field());
+                            }
```
I added tests. However, while the following works:

```
row wkt = "POINT(42.97109629958868 14.7552534006536)"
| STATS c = ST_CENTROID(TO_GEOPOINT(wkt));
```

Doing the string literal inline does not:

```
row dummy = "dummy"
| STATS c = ST_CENTROID(TO_GEOPOINT("POINT(42.97109629958868 14.7552534006536)"));
```

Yeilds:

```
org.elasticsearch.xpack.ql.InvalidArgumentException:
  Does not support yet aggregations over constants
  - [ST_CENTROID(TO_GEOPOINT("POINT(42.97109629958868 14.7552534006536)"))]
```

That exception is throw by code in `AbstractPhysicalOperationProviders` that includes the comment:

```
// Only count can now support literals - all the other aggs should be optimized away
```

So it seems this is not yet supported.

craigtaverner replied:
Anyway, use of fields that are reference fields, not literal fields, does not even reach the local optimization because no `FieldExtractExec` is produced, and therefor not optimized. I added a test to the `PhysicalOptimizationTests` to verify this.

craigtaverner commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/aggregate/SpatialAggregateFunction.java:
```diff
@@ -0,0 +1,51 @@
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ql.expression.function.aggregate;
```
Done

craigtaverner committed:
Added some more code review updates, including nested tests

craigtaverner committed:
Get nested functions working, if only from source values for now

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_point_doc_values

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizer.java:
```diff
@@ -427,4 +434,43 @@ public static boolean isPushableFieldAttribute(Expression exp, Predicate<FieldAt
+                    for (NamedExpression aggExpr : agg.aggregates()) {
+                        if (aggExpr instanceof Alias as && as.child() instanceof SpatialAggregateFunction af) {
+                            // Firstly we tell the aggregation function that is should expect doc-values instead of source values
+                            var changed = as.replaceChild(af.withDocValues());
```
I did try this quickly, but it did not work, and anyway I think we can do this as part of the improvement to support nested functions. Right now I got at least the current code to work with nested functions, but without the doc-values optimization.

craigtaverner committed:
Code review update

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/physical/AggregateExec.java:
```diff
@@ -60,6 +60,10 @@ public AggregateExec replaceChild(PhysicalPlan newChild) {
 
+    public AggregateExec replaceAggregates(List<? extends NamedExpression> aggregates) {
+        return new AggregateExec(source(), child(), groupings, aggregates, mode, estimatedRowSize);
+    }
```
OK. I've removed this utility method

craigtaverner committed:
Code review update

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/physical/FieldExtractExec.java:
```diff
@@ -13,17 +13,25 @@
     private final Attribute sourceAttribute;
+    private final Set<Attribute> forStats;
 
-    public FieldExtractExec(Source source, PhysicalPlan child, List<Attribute> attributesToExtract) {
```
I'm not sure what that has top do with compatibility, but I've re-ordered the constructors, if that was what you were asking for.

costin replied:
```java
 public FieldExtractExec(Source source, PhysicalPlan child, List<Attribute> attributesToExtract) {
    this(source, child, attributesToExtract, emptySet())
 }
```

It might be you already did that and I missed you moving the constructor around :)

craigtaverner commented on server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java:
```diff
@@ -649,6 +649,14 @@ public interface BlockLoaderContext {
+         * For example, spatial types use doc-values for stats, but read from source for search because doc-values are modified
+         * from the original.
+         */
+        boolean forStats();
```
What is confusing for me here is that we'll set this value to false by default, and only set it to true for spatial aggregations loaded from fields. That makes complete sense for the spatial field mappers block loading, but all other types will get a BlockLoaderContext with `preferDocValues==false`, while simultaneously ignoring that, and preferring to load doc-values.

The term `forStats` is not making any statement about whether that implies anything about doc-values versus source loading, and the decision to load differently is entirely local to the field mapper.

All other types can say 'you don't want to use this for stats, well both doc-values and source are equally good enough for non-stats, so I'll load the one I think is fastest'. But if we call it `preferDocValues==false`, then all the other non-spatial types are basically saying "I know I've been told *NOT* to prefer doc-values, but I'm going to ignore that because I magically know this is not relevant to me".

costin replied:
How about `useDocValues()`, `loadDocValues`, etc...? My point is `forStats` doesn't tell much.

astefan commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/spatial.csv-spec:
```diff
@@ -49,6 +87,136 @@ WIIT           | POINT(105.176060419161 -5.242566777132)   | Radin Inten II
+| SORT scalerank DESC
+;
+
+centroid:geo_point                            | count:long | scalerank:i
```
Not related to this specific test, but for a more wide range of value handlers some `null` values could be added to the "airport" data set.

craigtaverner replied:
Added more data with null values and have aggregation working on that. The centroid of null is null. The centroid of data including a few nulls is the centroid of the non-null values.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizer.java:
```diff
@@ -427,4 +429,49 @@ public static boolean isPushableFieldAttribute(Expression exp, Predicate<FieldAt
+    private static class SpatialDocValuesExtraction extends OptimizerRule<AggregateExec> {
+        @Override
+        protected PhysicalPlan rule(AggregateExec aggregate) {
+            var foundAttribute = new Holder<Attribute>(null);
```
I am not sure if this is to blame or not (a single Holder for attributes, while there may be more than one while searching). So, I tested with `FROM airports2 | stats centroid1=ST_CENTROID(location), centroid2=ST_CENTROID(location2)` where `location` and `location2` are `geo_point`s. This fails with

```
Caused by: java.lang.ClassCastException: class org.elasticsearch.compute.data.BytesRefArrayBlock cannot be cast to class org.elasticsearch.compute.data.LongBlock (org.elasticsearch.compute.data.BytesRefArrayBlock and org.elasticsearch.compute.data.LongBlock are in unnamed module of loader java.net.FactoryURLClassLoader @550fa96f)
        at org.elasticsearch.compute.aggregation.spatial.SpatialCentroidGeoPointDocValuesAggregatorFunction.addRawInput(SpatialCentroidGeoPointDocValuesAggregatorFunction.java:64)
        at org.elasticsearch.compute.aggregation.Aggregator.processPage(Aggregator.java:42)
        at org.elasticsearch.compute.operator.AggregationOperator.addInput(AggregationOperator.java:79)
        at org.elasticsearch.compute.operator.Driver.runSingleLoopIteration(Driver.java:214)
        at org.elasticsearch.compute.operator.Driver.run(Driver.java:139)
        at org.elasticsearch.compute.operator.Driver$1.doRun(Driver.java:327)
```

Also, not sure if this is supposed to work or not, or maybe return a more descriptive and user-friendly error message.

astefan replied:
I had a quick look at this and this seems to be the culprit.
This code works for me:

```
        protected PhysicalPlan rule(AggregateExec aggregate) {
            Set<FieldAttribute> fieldAttributes = new HashSet<>();

            PhysicalPlan plan = aggregate.transformDown(UnaryExec.class, exec -> {
...
                            if (af.field() instanceof FieldAttribute fieldAttribute) {
                                // We need to both mark the field to load differently, and change the spatial function to know to use it
                                fieldAttributes.add(fieldAttribute);
                                changedAggregates = true;
...
                if (exec instanceof FieldExtractExec fieldExtractExec && fieldAttributes.size() > 0) {
                    // Tell the field extractor that it should extract the field from doc-values instead of source values
                    if (fieldAttributes.retainAll(fieldExtractExec.attributesToExtract())) {
                        exec = fieldExtractExec.withForStats(fieldAttributes);
                    }
                }
                return exec;
...
```

craigtaverner replied:
This was a bug, and I changed the Holder to a Set, and wrote a test in `spatial.csv-spec` that tests this, as well as a PhysicalPlanner test to test the correct behaviour of the planner.

astefan reviewed (approved):
Other than the bug fix for [this](https://github.com/elastic/elasticsearch/pull/104218/files#r1457462532) and the corresponding test, it LGTM.

craigtaverner committed:
Added second location column to airports for wider testing

costin commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:
This test comes from a separate PR - looks like a git foobar.

craigtaverner replied:
I cannot find your comment in the file view, was this something that was fixed by merging main?

Or do you mean the spatial test I added to this file? I could delete that but thought a test of the logical plan for a spatial aggregation was not entirely without value.

craigtaverner committed:
Use second location in tests, including nulls

Includes a test fix for loading and converting nulls to encoded longs.

craigtaverner committed:
Fixed bug supporting multi spatial aggregations in the local node

The local physical planner only marked a single field for stats loading, but marked all spatial aggregations for stats loading, which load to only one aggregation getting the right data, while the rest would get the wrong data.

craigtaverner committed:
Renamed forStats to fieldExtractPreference for clarity

Now the planner decides whether to load data from doc-values. To remove the confusion of preferDocValues==false in the non-spatial cases, we use an ENUM with the default value of NONE, to make it clear we're leaving the choice up to the field type in all non-spatial cases.

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_point_doc_values

craigtaverner committed:
EsqlSpecIT was failing on very high precision centroids on different computers

This was not reproducible on the development machine, but CI machines were sufficiently different to lead to very tiny precision changes over very large Kahan summations. We fixed this by reducing the need for precision checks in clustered integration tests.

craigtaverner committed:
Delete docs/changelog/104218.yaml

craigtaverner committed:
Revert "Delete docs/changelog/104218.yaml"

This reverts commit 12c6980881bb838a16648bbd5ca7b87254a8a675.

craigtaverner committed:
Fixed changelog entry

elasticsearchmachine commented:
@craigtaverner according to this PR's labels, it should not have a changelog YAML, but I can't delete it because it is closed. Please either delete the changelog from the appropriate branch yourself, or adjust the labels.
